Fix some issues with focus scopes

grokys opened the pull request:
## What does the pull request do?

As part of the fix for #13325 I needed to do some basic refactoring to `FocusManager` to fix our broken 2015-era focus scope handling.

- Correctly focus outermost focus scope when a control was removed (previously didn't have enough context to do so, and so this feature only worked when hiding/disabling a control)
- Correctly clear the focused element in a scope when it is removed
- Store the focused element in an attached property instead of a dictionary like WPF: means we need to do a little less bookkeeping

`FocusManager` really needs a rewrite like the one started in #8392 - this PR however doesn't even attempt that, it just aims to fix the existing focus scope handling.

## Breaking Changes

Previously the reference assembly for `FocusManager` was exporting a compiler-generated enumerable class:

```csharp
[PrivateApi]
public class FocusManager : IFocusManager
{
    [CompilerGenerated]
    public sealed class <GetFocusScopeAncestors>d__18 : IEnumerable<IFocusScope>, IEnumerable, IEnumerator<IFocusScope>, IEnumerator, IDisposable
    {
        public IInputElement <>3__control;

        [DebuggerHidden]
        public <GetFocusScopeAncestors>d__18(int <>1__state);
    }
}
```

This class wasn't exposed by the API so couldn't be used, but is no longer being generated after this PR. Added a suppression for it,

## Fixed issues

Fixes #13325

grokys committed:
Remove focus hack from Popup.

grokys committed:
Added failing focus scope tests.

grokys committed:
Refactor focus scopes in FocusManager.

- Store focused element within a scope using an attached property (like WPF)
- Store current focus root so that focus can be restored to that root when a focused control or active focus scope is removed

Fixes #13325

grokys commented on src/Avalonia.Controls/Primitives/PopupRoot.cs:
```diff
@@ -77,7 +77,7 @@ public PopupRoot(TopLevel parent, IPopupImpl impl, IAvaloniaDependencyResolver?
         /// Gets the control that is hosting the popup root.
         /// </summary>
-        Visual? IHostedVisualTreeRoot.Host => VisualParent;
+        Visual? IHostedVisualTreeRoot.Host => Parent as Visual ?? ParentTopLevel;
```
@Gillibald this fixes what I think was an incorrect change in [one of your commits](https://github.com/AvaloniaUI/Avalonia/commit/6210018f3b5e98063252d2bf2933e1739ed17f85#diff-b9c372514271a3da80207200fb37c0f5e9ef566ec7bff8db21f9190c6a1540cbR80). I think the change in that commit was wrong because `VisualParent` should always be `null` for `PopupRoot` considering that it's a visual root.

jankrib commented on src/Avalonia.Base/Input/FocusManager.cs:
```diff
@@ -58,94 +52,59 @@ static FocusManager()
+                return true;
             }
-            else if (Current != null)
+            else if (_focusRoot?.GetValue(FocusedElementProperty) is { } restore && 
```
I find the _focusRoot concept a bit confusing. Wouldn't it be better to just stick to focus scopes like in the old implementation?

At least in theory there could be another focus scope between the cleared scope and the VisualRoot.

grokys replied:
This is just following [WPF's behavior](https://github.com/AvaloniaUI/Avalonia/issues/13325#issuecomment-1772655137) of focusing the outermost focus scope when a focused control is removed. The reason we need to store the focus root is that at the time the focused control is removed we may not have all the information we need to find that outermost scope.

jankrib reviewed (commented):
Nice😃 This is going t help a lot!

grokys committed:
Suppress API compat error.

This was being produced for a compiler-generated enumerable class that was erroneously being included in the reference assembly for `FocusManager`.

grokys committed:
Remove focus hack from ContextMenu.

And add failing test now that the hack is removed.

grokys committed:
Try to return a rooted host visual.

Fixes failing test from previous comment where focus wasn't restored when closing a context menu.

grokys committed:
Merge branch 'master' into fixes/13325-focus-root-scope

grokys commented:
I think this PR should be ready for review now.
